onnxruntime/core/framework/sparse_utils.cc

```diff
@@ -0,0 +1,510 @@
+}
+
+template <typename T>
+struct NotZero {
+  bool operator()(T v) const {
+    return v != T{0};
+  }
+};
+
+template <>
+struct NotZero<std::string> {
+  bool operator()(const std::string& s) const {
+    return !s.empty();
+  }
+};
+
+#if !defined(ORT_MINIMAL_BUILD)
+template <typename T, typename ValueRecorder>
+void ScanAndRecordCsr(gsl::span<const T> src_span, int64_t cols,
+                      std::vector<int64_t>& inner, std::vector<int64_t>& outer,
+                      ValueRecorder recorder) {
+  int64_t row = 0;
+  int64_t index = 0;
+  outer.push_back(0);
+  NotZero<T> not_zero;
+  for (const auto& v : src_span) {
+    auto cur_row = index / cols;
+    if (cur_row != row) {
+      outer.push_back(static_cast<int64_t>(inner.size()));
+      row = cur_row;
+    }
+    if (not_zero(v)) {
+      auto cur_col = index - cur_row * cols;
+      inner.push_back(cur_col);
+      recorder(v);
+    }
+    ++index;
+  }
+  outer.push_back(static_cast<int64_t>(inner.size()));
+}
+
+Status DenseTensorToSparseCsr(const DataTransferManager& data_manager, const Tensor& src,
+                              const AllocatorPtr& cpu_allocator, const AllocatorPtr& dst_allocator,
+                              SparseTensor& dst) {
+  const auto& src_dims = src.Shape().GetDims();
+  if (src_dims.size() != 2) {
+    return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT,
+                           "Currently do not support dims higher than 2 dimensions: ", src_dims.size());
+  }
+
+  const bool is_string = src.IsDataTypeString();
+
+  if (is_string && dst_allocator->Info().device.Type() != OrtDevice::CPU) {
+    return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT,
+                           "Unable to convert strings tensor to a sparse tensor that not on CPU");
+  }
+
+  const IDataTransfer* data_transfer = data_manager.GetDataTransfer(cpu_allocator->Info().device,
+                                                                    dst_allocator->Info().device);
+  ORT_RETURN_IF_NOT(data_transfer != nullptr, "Unable to find a data transfer for copying from device type: ",
+                    cpu_allocator->Info().device.Type(), " to device type: ", dst_allocator->Info().device.Type());
+
+  const auto element_size = src.DataType()->Size();
+  gsl::span<const uint8_t> src_span;
+  Tensor src_cpu;
+  if (src.Location().device.Type() != OrtDevice::CPU) {
+    Tensor t(src.DataType(), src.Shape(), cpu_allocator);
+    ORT_RETURN_IF_ERROR(data_manager.CopyTensor(src, t));
+    src_cpu = std::move(t);
+    src_span = gsl::make_span(reinterpret_cast<const uint8_t*>(src_cpu.DataRaw()), src_cpu.SizeInBytes());
+  } else {
+    src_span = gsl::make_span(reinterpret_cast<const uint8_t*>(src.DataRaw()), src.SizeInBytes());
+  }
+
+  const auto rows = src_dims[0];
+  const auto cols = src_dims[1];
+
+  std::vector<int64_t> inner_indices;
+  std::vector<int64_t> outer_indices;
```

Comment:
Can we do a reserve for inner_indices? Even if it's a rough guess I would expect it to be better than gradually growing from empty.

Reply:
I r emoved the reservation because I realized I do not know what the rough guess might be. It can be 0. But if you have some value in mind, I'd put reserve back in again.